File: 22-GenerateParentheses/22-GenerateParentheses.py
# Last updated: 6/11/2026, 11:38:27 AM
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        ans = []
        def rec(path, open, close):
            if(open>n):
                return
            if(close>open):
                return
            if(len(path)==2*n):
                ans.append(path)
                return
            rec(path+"(", open+1, close)
            rec(path+")", open, close+1)
        rec("", 0, 0)
        return ans

File: 22-GenerateParentheses/22-GenerateParentheses.cpp
// Last updated: 6/11/2026, 11:38:20 AM
class Solution {
public:
    vector<string> all;
    int n;
    void rec(string &s, int open){
        int close = s.size() - open;
        if(close>open) return;
        if(s.size() == n){
            if(open == close){
                all.push_back(s);
                cout<<s<<endl;
            }
            return;
        }
        s.push_back('(');
        rec(s, open+1);
        s.pop_back();

        s.push_back(')');
        rec(s, open);
        s.pop_back();
    }
    vector<string> generateParenthesis(int n) {
        this->n = 2*n;
        string s = "";
        rec(s, 0);
        return all;
    }
};